### PR TITLE
classe webcam

### PR DESCRIPTION
**Pull request by Elvisnando**

### CODE REVIEW

File: lib/webcam.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#ifndef WEBCAM_H
+#define WEBCAM_H
+
+#include <libudev.h>
+#include <QtCore/QString>
+
+#include "device.h"
+#include <libudev.h>
```
Comment (nwoki) [inline lib/webcam.h]: double include

File: lib/device.h [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #include "udevnotifier_global.h"
 
 #include <QtCore/QMetaType>
+#include <QtCore/QThread>
```
Comment (nwoki) [inline lib/device.h]: why?

Comment (nwoki) [inline lib/device.h]: why?

File: lib/device.h [hunk 3]
```diff
@@ -23,17 +24,28 @@ class DevicePrivate;
 class UDEVNOTIFIERSHARED_EXPORT Device
 {
 public:
+
+    enum TIPO{
```
Comment (nwoki) [inline lib/device.h]: eng please

File: lib/device.h [hunk 4]
```diff
@@ -23,17 +24,28 @@ class DevicePrivate;
...
+
+        MONITOR,
+        WEBCAM
```
Comment (nwoki) [inline lib/device.h]: what happens if it's not any of the above?

File: lib/device.h [hunk 5]
```diff
@@ -23,17 +24,28 @@ class DevicePrivate;
...
+
+
     Device(udev_device *device = nullptr);
     Device(const Device &other);
     ~Device();
 
     QString node() const;
     QString subsystem() const;
     QString type() const;
+    TIPO   Typedevice();
```
Comment (nwoki) [inline lib/device.h]: "camelCase" please

File: lib/device.h [hunk 6]
```diff
@@ -23,17 +24,28 @@ class DevicePrivate;
...
 
 private:
     DevicePrivate * const d;
 
+
```
Comment (nwoki) [inline lib/device.h]: extra space for nothing

File: lib/lib.pro [hunk 7]
```diff
@@ -18,17 +18,22 @@ DEFINES += QT_DEPRECATED_WARNINGS
...
 
 HEADERS += \
     device.h \
     device_p.h \
-    monitor.h  \
     udevnotifier.h \
     udevnotifier_global.h \
-    udevnotifier_p.h
+    udevnotifier_p.h \
+    webcam.h\
```
Comment (nwoki) [inline lib/lib.pro]: alphabetical order

File: lib/monitor.cpp [hunk 8]
```diff
@@ -1,16 +1,22 @@
...
+#include <QtCore/QDebug>
 
 namespace UdevNotifier {
 
 Monitor::Monitor(udev_device *device)
     :Device(device)
 {
+    qDebug() << "qyesto è un monitr";
+
+
```
Comment (nwoki) [inline lib/monitor.cpp]: extra space for nothing

File: lib/monitor.cpp [hunk 9]
```diff
@@ -1,16 +1,22 @@
...
 #include "device.h"
-
+#include <QtCore/QDebug>
 
 namespace UdevNotifier {
 
 Monitor::Monitor(udev_device *device)
     :Device(device)
 {
+    qDebug() << "qyesto è un monitr";
```
Comment (nwoki) [inline lib/monitor.cpp]: don't need this

File: lib/monitor.cpp [hunk 10]
```diff
@@ -1,16 +1,22 @@
 #include "monitor.h"
 #include "device.h"
-
+#include <QtCore/QDebug>
```
Comment (nwoki) [inline lib/monitor.cpp]: don't need

File: lib/monitor.cpp [hunk 11]
```diff
@@ -1,16 +1,22 @@
...
     :Device(device)
 {
+    qDebug() << "qyesto è un monitr";
+
+
 }
 
 Monitor::~Monitor()
 {
+
```
Comment (nwoki) [inline lib/monitor.cpp]: extra space for nothing

File: lib/device.h (hunk 2 above)
Comment (Elvisnando) [reply lib/device.h]: era solo una prova elimino

File: lib/webcam.cpp [hunk 12]
```diff
@@ -0,0 +1,16 @@
+#include "webcam.h"
+#include "device.h"
+#include "device_p.h"
+#include <QtCore/QDebug>
+namespace UdevNotifier {
+
+Webcam::Webcam(udev_device *device):Device(device)
+{
+    qDebug()<<"Questa è una web cam";
```
Comment (nwoki) [inline lib/webcam.cpp]: do not need

File: lib/webcam.cpp [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#include "webcam.h"
+#include "device.h"
+#include "device_p.h"
+#include <QtCore/QDebug>
```
Comment (nwoki) [inline lib/webcam.cpp:4]: do not need

File: lib/webcam.h [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+
+
+class UDEVNOTIFIERSHARED_EXPORT Webcam: public Device
+{
+public:
+    Webcam(udev_device *device = nullptr);
+    ~Webcam();
+
+
```
Comment (nwoki) [inline lib/webcam.h]: extra space for nothing

File: lib/device.cpp [hunk 15]
```diff
@@ -67,7 +69,26 @@ QString Device::type() const
     return d->type;
 }
 
+Device::TIPO Device::Typedevice(){
+
+ if(d->type=="drm_minor"){
+    return MONITOR;
+    }
 
+    if(d->subsystem=="video4linux"){
```
Comment (nwoki) [inline lib/device.cpp]: why use type on one if and subsystem on the other?

Comment (Elvisnando) [reply lib/device.cpp]: perche il in tybe mi dice che è un usb   generico con il susystem rilevo  video4linux

File: lib/device.h (hunk 4 above)
Comment (Elvisnando) [reply lib/device.h]: ritorna -256 come valore

File: lib/device.cpp (hunk 15 above)
Comment (nwoki) [reply lib/device.cpp]: * fix code indentation

File: lib/device.h [hunk 16]
```diff
@@ -23,17 +21,24 @@ class DevicePrivate;
 class UDEVNOTIFIERSHARED_EXPORT Device
 {
 public:
+
+    enum TYPE{
+
```
Comment (nwoki) [inline lib/device.h]: again: what if it's none of these two types?

File: lib/device.h [hunk 17]
```diff
@@ -23,17 +21,24 @@ class DevicePrivate;
...
+        MONITOR,
+        WEBCAM
+    };
+
     Device(udev_device *device = nullptr);
     Device(const Device &other);
     ~Device();
 
+    TYPE   Typedevice();
```
Comment (nwoki) [inline lib/device.h]: camelCase pls

File: lib/device.cpp [hunk 18]
```diff
@@ -67,7 +69,26 @@ QString Device::type() const
...
+ }
 
 }
 
+
+
+
+
+
+
```
Comment (nwoki) [inline lib/device.cpp]: extra space for nothing

File: lib/device.h [hunk 19]
```diff
@@ -22,21 +22,22 @@ class UDEVNOTIFIERSHARED_EXPORT Device
 {
 public:
 
-    enum TYPE{
-
+    enum TYPE {
```
Comment (nwoki) [inline lib/device.h]: DeviceType

File: lib/device.h [hunk 20]
```diff
@@ -22,21 +22,22 @@ class UDEVNOTIFIERSHARED_EXPORT Device
...
         MONITOR,
         WEBCAM,
         USBDEVICE
     };
 
-    Device(udev_device *device = nullptr, TYPE n  = USBDEVICE);
+    Device(udev_device *device = nullptr, TYPE typeD  = USBDEVICE);
```
Comment (nwoki) [inline lib/device.h]: deviceType

File: lib/device.h [hunk 21]
```diff
@@ -22,21 +22,22 @@ class UDEVNOTIFIERSHARED_EXPORT Device
...
     Device(const Device &other);
     ~Device();
 
     QString node() const;
     QString subsystem() const;
     QString type() const;
-    TYPE d_typD() const;
+    TYPE tDevice();
```
Comment (nwoki) [inline lib/device.h]: deviceType

File: lib/device.h [hunk 22]
```diff
@@ -22,21 +22,22 @@ class UDEVNOTIFIERSHARED_EXPORT Device
...
+    TYPE typeDevice;
```
Comment (nwoki) [inline lib/device.h]: deviceType

File: lib/monitor.cpp [hunk 23]
```diff
@@ -3,10 +3,9 @@
 
 namespace UdevNotifier {
 
-Monitor::Monitor(udev_device *device, TYPE n)
-    :Device(device,n)
+Monitor::Monitor(udev_device *device)
+    :Device(device,MONITOR)
```
Comment (nwoki) [inline lib/monitor.cpp]: spacing

File: lib/webcam.cpp [hunk 24]
```diff
@@ -4,9 +4,9 @@
 #include <QtCore/QDebug>
 namespace UdevNotifier {
 
-Webcam::Webcam(udev_device *device,TYPE typeD):Device(device,typeD)
+Webcam::Webcam(udev_device *device)
+    :Device(device,WEBCAM)
```
Comment (nwoki) [inline lib/webcam.cpp]: spacing

File: lib/device.cpp [hunk 25]
```diff
@@ -9,9 +9,8 @@ namespace UdevNotifier {
 Device::Device(udev_device* device, TYPE typeD)
     : d(new DevicePrivate)
 {
-
-    d->d_typD=typeD;
     d->device = device;
+    this->typeDevice=typeD;
```
Comment (nwoki) [inline lib/device.cpp]: this is implicit

File: lib/device.cpp [hunk 26]
```diff
@@ -21,8 +20,7 @@ Device::Device(udev_device* device, TYPE typeD)
 
 
       //  qDebug() << QString("%1 - %2 - %3").arg(d->node).arg(d->subsystem).arg(d->type);
-     //   qDebug()<< typeDevice();
-        qDebug() << "dal d pointer"<<d->d_typD;
+        qDebug() << "TIPO :" << tDevice();
```
Comment (nwoki) [inline lib/device.cpp]: change

File: lib/device.cpp [hunk 27]
```diff
@@ -69,12 +66,10 @@ QString Device::type() const
     return d->type;
 }
 
-Device::TYPE Device::d_typD() const {
-return d->d_typD;
-
+Device::TYPE Device::tDevice()
```
Comment (nwoki) [inline lib/device.cpp]: change -> deviceType and make is a const reference

File: lib/device.cpp [hunk 28]
```diff
@@ -69,12 +66,10 @@ QString Device::type() const
...
+{
+return typeDevice;
```
Comment (nwoki) [inline lib/device.cpp]: indent

File: lib/device.cpp [hunk 29]
```diff
@@ -6,11 +6,11 @@
 namespace UdevNotifier {
 
 
-Device::Device(udev_device* device, TYPE typeD)
+Device::Device(udev_device* device, TYPEDEVICE typeD)
```
Comment (nwoki) [inline lib/device.cpp]: DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType

File: lib/device.cpp [hunk 30]
```diff
@@ -6,11 +6,11 @@
...
     : d(new DevicePrivate)
 {
     d->device = device;
-    this->typeDevice=typeD;
+    typeDevice=typeD;
```
Comment (nwoki) [inline lib/device.cpp]: spacing

File: lib/device.h [hunk 31]
```diff
@@ -22,21 +22,21 @@ class UDEVNOTIFIERSHARED_EXPORT Device
 {
 public:
 
-    enum TYPE {
+    enum TYPEDEVICE {
```
Comment (nwoki) [inline lib/device.h]: DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType DeviceType

File: lib/device.cpp [hunk 32]
```diff
@@ -1,18 +1,16 @@
...
 #include <libudev.h>
-
 #include <QtCore/QDebug>
 
-
 namespace UdevNotifier {
 
 
-Device::Device(udev_device* device)
+Device::Device(udev_device* device, DeviceType deviceT )
```
Comment (nwoki) [inline lib/device.cpp]: deviceT -> deviceType

File: lib/device.cpp [hunk 33]
```diff
@@ -1,18 +1,16 @@
...
     : d(new DevicePrivate)
 {
     d->device = device;
+    typeDevice = deviceT;
```
Comment (nwoki) [inline lib/device.cpp]: same here. deviceType

File: lib/device.cpp [hunk 34]
```diff
@@ -22,6 +20,7 @@ Device::Device(udev_device* device)
 
 
       //  qDebug() << QString("%1 - %2 - %3").arg(d->node).arg(d->subsystem).arg(d->type);
+        qDebug() << "TIPO :" << deviceType();
```
Comment (nwoki) [inline lib/device.cpp]: no italian. Only eng

File: lib/device.cpp [hunk 35]
```diff
@@ -34,6 +33,7 @@ Device::Device(udev_device* device)
 
 
         qDebug()<<"**************************************************************";
+
```
Comment (nwoki) [inline lib/device.cpp]: remove whitespace

File: lib/device.cpp [hunk 36]
```diff
@@ -67,7 +66,10 @@ QString Device::type() const
     return d->type;
 }
 
-
+Device::DeviceType Device::deviceType() const
+{
+    return typeDevice;
```
Comment (nwoki) [inline lib/device.cpp]: var name

File: lib/device.h [hunk 37]
```diff
@@ -23,17 +21,26 @@ class DevicePrivate;
...
 public:
-    Device(udev_device *device = nullptr);
+
+    enum DeviceType {
+        MONITOR,
+        WEBCAM,
+        USBDEVICE
+    };
+
+    Device(udev_device *device = nullptr, DeviceType deviceType  = USBDEVICE);
```
Comment (nwoki) [inline lib/device.h]: careful, extra space here

File: lib/device.h [hunk 38]
```diff
@@ -23,17 +21,26 @@ class DevicePrivate;
...
     Device(const Device &other);
     ~Device();
 
     QString node() const;
     QString subsystem() const;
     QString type() const;
+    DeviceType deviceType() const;
+    DeviceType typeDevice;
```
Comment (nwoki) [inline lib/device.h]: device type

File: lib/device_p.h [hunk 39]
```diff
@@ -3,6 +3,7 @@
 
 #include <QtCore/QString>
 
+
```
Comment (nwoki) [inline lib/device_p.h]: extra space not needed

File: lib/udevnotifier.cpp [hunk 40]
```diff
@@ -5,7 +5,7 @@
 #include <libudev.h>
 #include <monitor.h>
 #include <poll.h>
-
+#include"webcam.h"
```
Comment (nwoki) [inline lib/udevnotifier.cpp:8]: space

File: lib/udevnotifier.cpp [hunk 41]
```diff
@@ -90,10 +90,10 @@ void UdevNotifier::run()
                 Q_EMIT udevEvent(actionFromString(udev_device_get_action(dev)), new Monitor(dev));
             }
 
-            if (udev_device_get_devtype(dev) == QStringLiteral("usb_device")) {
-                Q_EMIT udevEvent(actionFromString(udev_device_get_action(dev)), new Device(dev));
+            if (udev_device_get_subsystem(dev) == QStringLiteral("video4linux")) {
+                Q_EMIT udevEvent(actionFromString(udev_device_get_action(dev)), new Webcam(dev));
             }
-
+          //  Q_EMIT udevEvent(actionFromString(udev_device_get_action(dev)), new Device(dev));
```
Comment (nwoki) [inline lib/udevnotifier.cpp]: if you don't use it, remove it

File: lib/device.cpp (hunk 32 above)
Comment (Elvisnando) [reply lib/device.cpp]: ok, but name of function "deviceType" become "deviceTypeReturn"

File: lib/device.cpp [hunk 42]
```diff
@@ -20,19 +18,16 @@ Device::Device(udev_device* device)
         d->subsystem = udev_device_get_subsystem(d->device);
         d->type = udev_device_get_devtype(d->device);
 
-
-      //  qDebug() << QString("%1 - %2 - %3").arg(d->node).arg(d->subsystem).arg(d->type);
+        // qDebug() << QString("%1 - %2 - %3").arg(d->node).arg(d->subsystem).arg(d->type);
+        qDebug() << "TYPE :" << Device::deviceType(); // Test funtion
```
Comment (nwoki) [inline lib/device.cpp:22]: no need to specify "Device::". The base class function will be automatically called

File: lib/device.h [hunk 43]
```diff
@@ -23,17 +21,25 @@ class DevicePrivate;
 class UDEVNOTIFIERSHARED_EXPORT Device
 {
 public:
-    Device(udev_device *device = nullptr);
+
+    enum DataType {
```
Comment (nwoki) [inline lib/device.h:25]: it's not a DataType, but a DeviceType

**Review by nwoki — CHANGES_REQUESTED** [review]
Almost there. Fix the comments i added and then you're good to go

**Review by nwoki — APPROVED** [review]

**Pull request merged**